sql.wit.md

```diff
@@ -0,0 +1,71 @@
+# `wasi-sql` API
+
+## Interfaces
+
+```wit
+interface "wasi:sql" {
+    // iterator item type
+    record item {
+        field-name: string
+        values: data-type
+        index: u32
+    }
+    
+    // common data types
+    variant data-type {
+        int32(s32),
+        int64(s64),
+        uint32(u32),
+        uint64(u64),
+        float(float64),
+        double(float64),
+        str(string),
+        boolean(bool),
+        date(string),
+        time(string),
+        timestamp(string),
+        binary(list<u8>),
+        null
+    }
+    
+    // iterator for item
+    resource row {
+        next: func() -> option<row>
+    }
+    
+    // allows parameterized queries
+    resource statement {
+        // e.g., create("SELECT * FROM users WHERE name = ? AND age = ?", vec!["John Doe", "32"])
+        create: func(s: string, p: list<string>) -> result<statement, error>
+    }
+    
+    // query is optimized for querying data, and 
+    // implementors can make use of that fact to optimize 
+    // the performance of query execution (e.g., using
+    // indexes).
+    query: func(q: statement) -> result<row, error>
```

Comment:
I would expect the `query()` to be able to return multiple rows. Is there a plan to add a new resource as collection of rows?

Reply:
It looks at the moment like `query` is returning the first row, and then you call `next` to get to the next row - this fits with the sample in the read-me.  But then I can't see how to get to an `item`.  @danbugs I think something might have got out of whack during the various revisions - looking at the read-me has `row` gotten mixed up with `item` in some places?

FWIW, the Spin prototype has `query` returning a `rowset` which contained a list of `row`, and a `row` contained a list of `value` (the field info was separate).  However that doesn't handle very large datasets well (everything has to be realised into the list).  This might be remedies by streams, I'm not sure; but in the meantime, a resource that fetches the next record on demand (potentially lazily) makes sense to me.  Not sure if there are performance considerations relating to going across a marshalling boundary on every row though.

Reply:
>  looking at the read-me has row gotten mixed up with item in some places?

It made me confused hence my comments.

>  This might be remedies by streams

I assume that `streams` would be the right solution here at some point, might be wrong though.

> a resource that fetches the next record on demand (potentially lazily) makes sense to me

That would my expectation here as well.

Reply:
I'm wondering about how to get the item as well. I also wonder, does `query()` block until a row is ready, or does it return immediately (possibly with an error indicating the query couldn't be run) and calling `row.next()` blocks until the row is ready? I'm just wondering if it will be possible to fire queries on multiple connections in the case where the queries are long running and you can allow to server to process them simultaneously.